src/IO/writers.jl

```diff
@@ -0,0 +1,391 @@
+    file = h5open(filename, "w")
+    cache = Dict{String, String}()
+    return HDF5Writer(file, cache)
+end
+
+function cartesianindices_to_matrix(elemorder)
+    m, n = length(elemorder), length(eltype(elemorder))
+    elemordermatrix = zeros(Int, m, n)
+    for (i, order) in enumerate(elemorder)
+        for j in 1:n
+            elemordermatrix[i, j] = order[j]
+        end
+    end
+    return elemordermatrix
+end
+
+# write_new!(writer, obj) should return the name of the written object
+# write(writer, obj) checks the cache: if the object has not yet been written it then calls write_new!
+
+"""
+    write!(reader, obj[, path])
+
+Checks the cache to verify if the object, specified by the `path`, is in the cache.
+If it has not been read, this function reads it to the cache and returns the name of the object.
+"""
+write!(writer::HDF5Writer, obj, name = defaultname(obj)) =
+    get!(writer.cache, name) do
+        write_new!(writer, obj, name)
+    end
+
+# Domains
+defaultname(::Domains.SphereDomain) = "domains/sphere"
+defaultname(::Domains.IntervalDomain) = "domains/interval"
+
+"""
+    write_new!(writer, domain, name)
+
+Writes an object of type 'IntervalDomain' and name 'name' to the HDF5 file.
+"""
+function write_new!(
+    writer::HDF5Writer,
+    domain::Domains.IntervalDomain,
+    name::AbstractString = defaultname(domain),
+)
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalDomain")
+    write_attribute(group, "coord_type", string(typeof(domain.coord_min)))
+    write_attribute(group, "coord_min", string(domain.coord_min))
+    write_attribute(group, "coord_max", string(domain.coord_max))
+    !isnothing(domain.boundary_names) &&
+        write_attribute(group, "boundary_names", string(domain.boundary_names))
+    return name
+end
+
+"""
+    write_new!(writer, domain, name)
+
+Writes an object of type 'SphereDomain' and name 'name' to the HDF5 file.
+"""
+function write_new!(
+    writer::HDF5Writer,
+    domain::Domains.SphereDomain,
+    name::AbstractString = defaultname(domain),
+)
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "SphereDomain")
+    write_attribute(group, "radius", domain.radius)
+    return name
+end
+
+# Meshes
+defaultname(::Meshes.IntervalMesh) = "meshes/intervalmesh"
+defaultname(::Meshes.RectilinearMesh) = "meshes/mesh2d"
+defaultname(::Meshes.AbstractCubedSphere) = "meshes/mesh2d"
+
+"""
+    write_new!(writer, mesh, name)
+
+Write `IntervalMesh` data to HDF5.
+
+"""
+function write_new!(
+    writer::HDF5Writer,
+    mesh::Meshes.IntervalMesh,
+    name::AbstractString = defaultname(mesh),
+)
+    meshname = split(name, "/")[end]
+    group = create_group(writer.file, name)
+    write_attribute(group, "type", "IntervalMesh")
+    write!(writer, mesh.domain, "domains/$meshname")
+    write_attribute(group, "nelements", length(mesh.faces))
+    write_attribute(group, "faces", string(mesh.faces))
```

Comment:
if `faces` is an `Array`, it should be written as an array. If it's a range, we shouldn't need to write it?

Reply:
`faces` can be range. We we like it to be written as an `Array` anyways?

Reply:
Done. `Range` is used to construct a uniform mesh.